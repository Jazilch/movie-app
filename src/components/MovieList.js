/* eslint-disable camelcase */

'use es6';

import React, { Component } from 'react';
import styled from 'styled-components/macro';
import MovieItem from './MovieItem';
import DefaultLoader from './DefaultLoader';
import MovieListToolbar from './MovieListToolbar';

const MovieListGrid = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-column-gap: 10px;
  grid-row-gap: 10px;
  @media (max-width: 950px) {
    grid-template-columns: repeat(2, 1fr);
  }
  @media (max-width: 767px) {
    grid-template-columns: repeat(1, 1fr);
  }
`;

class MovieList extends Component {
  componentDidMount() {
    const { getMovies } = this.props;
    getMovies();
  }

  showMoviePage = id => () => {
    const { history } = this.props;
    history.push(`/movies/${id}`);
  };

  filterBySearch = movies => {
    const { searchTerm } = this.props;
    return movies.title
      .toLowerCase()
      .includes(String(searchTerm).toLowerCase());
  };

  getFilteredMedia = movies => {
    return movies.filter(this.filterBySearch);
  };

  render() {
    const { movies } = this.props;
    if (!movies.data) return <DefaultLoader />;
    const filteredMovies = this.getFilteredMedia(movies.data);
    return (
      <>
        <MovieListToolbar />
        <MovieListGrid>
          {filteredMovies.map(({ id, title, overview, poster_path }) => {
            return (
              <MovieItem
                key={id}
                id={id}
                title={title}
                overview={overview}
                poster_path={poster_path}
                showMoviePage={this.showMoviePage}
              />
            );
          })}
        </MovieListGrid>
      </>
    );
  }
}

export default MovieList;
