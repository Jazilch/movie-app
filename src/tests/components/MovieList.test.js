/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable no-shadow */
/* eslint-disable no-undef */

'use es6';

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MovieList from '../../components/MovieList';

const movies = {
  data: [
    {
      id: 1,
      title: 'Joker',
      overview: 'Description of the Joker',
      genres: ['Crime', 'Drama', 'Thriller']
    },
    { id: 2, title: 'Terminator', overview: 'I am Arnold', genres: ['Action'] },
    {
      id: 3,
      title: 'Ready or Not',
      overview: "A bride's wedding night ",
      genres: ['Comedy', 'Horror', 'Mystery', 'Thriller']
    }
  ]
};

describe('MovieList component', () => {
  let movielist;
  let getMoviesMock;
  let store;
  beforeEach(() => {
    getMoviesMock = jest.fn();
    store = {
      movies,
      searchTerm: '',
      filterTerm: 'All',
      getMovies: getMoviesMock
    };
    movielist = shallow(<MovieList {...store} />);
  });

  it('renders component correctly', () => {
    expect(toJson(movielist)).toMatchSnapshot();
  });

  it('calls ComponentDidMount once', () => {
    const componentDidMountSpy = jest.spyOn(
      MovieList.prototype,
      'componentDidMount'
    );
    shallow(<MovieList {...store} />);
    expect(componentDidMountSpy).toHaveBeenCalled();
    componentDidMountSpy.mockReset();
    componentDidMountSpy.mockRestore();
  });

  it('renders a loading indicator if no data', () => {
    movielist.setProps({ movies: { data: null } });
    expect(movielist.find('DefaultLoader').exists()).toEqual(true);
  });

  it('shows the correct number of items', () => {
    expect(movielist.find('MovieList__MovieListGrid').children().length).toBe(
      3
    );
  });

  it('can search movies via a search term', () => {
    movielist.setProps({ searchTerm: 'Term' });
    expect(
      movielist
        .find('MovieList__MovieListGrid')
        .children()
        .prop('title')
    ).toEqual('Terminator');
  });

  it('is case insensitive', () => {
    movielist.setProps({ searchTerm: 'term' });
    expect(
      movielist
        .find('MovieList__MovieListGrid')
        .children()
        .prop('title')
    ).toEqual('Terminator');
  });

  it('can filter results', () => {
    movielist.setProps({ filterTerm: 'Horror' });
    expect(
      movielist
        .find('MovieList__MovieListGrid')
        .children()
        .prop('title')
    ).toEqual('Ready or Not');
  });

  it('shows empty state if no results', () => {
    movielist.setProps({ filterTerm: 'Western' });
    expect(movielist.find('SearchEmptyState').exists()).toEqual(true);
  });
});
