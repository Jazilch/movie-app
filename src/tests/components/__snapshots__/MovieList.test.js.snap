// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component renders component correctly 1`] = `
<Fragment>
  <MovieListToolbar />
  <MovieList__MovieListGrid>
    <MovieItem
      genres={
        Array [
          "Crime",
          "Drama",
          "Thriller",
        ]
      }
      id={1}
      key="1"
      overview="Description of the Joker"
      showMoviePage={[Function]}
      title="Joker"
    />
    <MovieItem
      genres={
        Array [
          "Action",
        ]
      }
      id={2}
      key="2"
      overview="I am Arnold"
      showMoviePage={[Function]}
      title="Terminator"
    />
    <MovieItem
      genres={
        Array [
          "Comedy",
          "Horror",
          "Mystery",
          "Thriller",
        ]
      }
      id={3}
      key="3"
      overview="A bride's wedding night "
      showMoviePage={[Function]}
      title="Ready or Not"
    />
  </MovieList__MovieListGrid>
</Fragment>
`;
